File: kover-features-jvm/src/main/java/kotlinx/kover/features/jvm/LegacyVerification.java
/*
 * Copyright 2017-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
 */

package kotlinx.kover.features.jvm;

import com.intellij.rt.coverage.aggregate.api.AggregatorApi;
import com.intellij.rt.coverage.aggregate.api.Request;
import com.intellij.rt.coverage.report.api.Filters;
import com.intellij.rt.coverage.verify.Verifier;
import com.intellij.rt.coverage.verify.api.*;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.TreeMap;

class LegacyVerification {
    private LegacyVerification() {
        // no-op
    }

    static final BigDecimal ONE_HUNDRED = new BigDecimal(100);

    public static List<KoverLegacyFeatures.RuleViolations> verify(List<KoverLegacyFeatures.Rule> rules, File tempDir, KoverLegacyFeatures.ClassFilters filters, List<File> reports, List<File> outputs) throws IOException {
        Filters intellijFilters = Wildcards.convertFilters(filters);

        final ArrayList<Rule> rulesArray = new ArrayList<>();

        File ic = new File(tempDir, "agg-ic.ic");
        File smap = new File(tempDir, "agg-smap.smap");

        Request requests = new Request(intellijFilters, ic, smap);
        AggregatorApi.aggregate(Collections.singletonList(requests), reports, outputs);

        for (int ruleIndex = 0; ruleIndex < rules.size(); ruleIndex++) {
            KoverLegacyFeatures.Rule rule = rules.get(ruleIndex);

            final List<Bound> bounds = new ArrayList<>();
            for (int boundIndex = 0; boundIndex < rule.bounds.size(); boundIndex++) {
                KoverLegacyFeatures.Bound b = rule.bounds.get(boundIndex);

                bounds.add(new Bound(boundIndex, counterToIntellij(b), valueTypeToIntellij(b), valueToIntellij(b, b.minValue), valueToIntellij(b, b.maxValue)));
            }

            rulesArray.add(new Rule(ruleIndex, ic, targetToIntellij(rule), bounds));
        }


        final Verifier verifier = new Verifier(rulesArray);
        verifier.processRules();
        List<RuleViolation> violations = VerificationApi.verify(rulesArray);

        ArrayList<KoverLegacyFeatures.RuleViolations> ruleViolations = new ArrayList<>();
        for (RuleViolation ruleViolation : violations) {
            TreeMap<ViolationId, KoverLegacyFeatures.BoundViolation> resultBounds = new TreeMap<>();

            KoverLegacyFeatures.Rule rule = rules.get(ruleViolation.id);
            for (int boundIndex = 0; boundIndex < ruleViolation.violations.size(); boundIndex++) {
                BoundViolation boundViolation = ruleViolation.violations.get(boundIndex);
                KoverLegacyFeatures.Bound bound = rule.bounds.get(boundViolation.id);

                for (Violation maxViolation : boundViolation.maxViolations) {
                    String entityName = rule.groupBy != KoverLegacyFeatures.GroupingBy.APPLICATION ? maxViolation.targetName : null;
                    resultBounds.put(
                            new ViolationId(boundViolation.id, entityName),
                            new KoverLegacyFeatures.BoundViolation(bound, true, intellijToValue(maxViolation.targetValue, bound), entityName)
                    );
                }
                for (Violation minViolation : boundViolation.minViolations) {
                    String entityName = rule.groupBy != KoverLegacyFeatures.GroupingBy.APPLICATION ? minViolation.targetName : null;
                    resultBounds.put(
                            new ViolationId(boundViolation.id, entityName),
                            new KoverLegacyFeatures.BoundViolation(bound, false, intellijToValue(minViolation.targetValue, bound), entityName)
                    );
                }
            }

            ruleViolations.add(new KoverLegacyFeatures.RuleViolations(rule, new ArrayList<>(resultBounds.values())));
        }

        return ruleViolations;
    }

    private static class ViolationId implements Comparable<ViolationId> {
        private final int index;
        private final String entityName;

        private ViolationId(int index, String entityName) {
            this.index = index;
            this.entityName = entityName;
        }

        @Override
        public int compareTo(@NotNull LegacyVerification.ViolationId other) {
            // first compared by index
            if (index != other.index) {
                return Integer.compare(index, other.index);
            }

            // if indexes are equals then compare by entity name
            if (entityName == null) {
                // bounds with empty entity names goes first
                return (other.entityName == null) ? 0 : -1;
            }

            if (other.entityName == null) return 1;

            return entityName.compareTo(other.entityName);
        }
    }

    private static BigDecimal intellijToValue(BigDecimal intellijValue, KoverLegacyFeatures.Bound bound) {
        if (isPercentage(bound.aggregationForGroup)) {
            return intellijValue.multiply(ONE_HUNDRED);
        } else {
            return intellijValue;
        }
    }

    private static Target targetToIntellij(KoverLegacyFeatures.Rule rule) {
        switch (rule.groupBy) {
            case APPLICATION:
                return Target.ALL;
            case CLASS:
                return Target.CLASS;
            case PACKAGE:
                return Target.PACKAGE;
        }
        return null;
    }

    private static Counter counterToIntellij(KoverLegacyFeatures.Bound bound) {
        switch (bound.coverageUnits) {
            case LINE:
                return Counter.LINE;
            case INSTRUCTION:
                return Counter.INSTRUCTION;
            case BRANCH:
                return Counter.BRANCH;
        }
        return null;
    }

    private static ValueType valueTypeToIntellij(KoverLegacyFeatures.Bound bound) {
        switch (bound.aggregationForGroup) {
            case COVERED_COUNT:
                return ValueType.COVERED;
            case MISSED_COUNT:
                return ValueType.MISSED;
            case COVERED_PERCENTAGE:
                return ValueType.COVERED_RATE;
            case MISSED_PERCENTAGE:
                return ValueType.MISSED_RATE;
        }
        return null;
    }

    private static BigDecimal valueToIntellij(KoverLegacyFeatures.Bound bound, BigDecimal value) {
        if (value == null) return null;

        if (isPercentage(bound.aggregationForGroup)) {
            return value.divide(ONE_HUNDRED, 6, RoundingMode.HALF_UP);
        } else {
            return value;
        }
    }

    private static boolean isPercentage(KoverLegacyFeatures.AggregationType aggregationType) {
        return aggregationType == KoverLegacyFeatures.AggregationType.COVERED_PERCENTAGE || aggregationType == KoverLegacyFeatures.AggregationType.MISSED_PERCENTAGE;
    }
}

File: kover-features-jvm/src/main/java/kotlinx/kover/features/jvm/impl/LegacyVerification.kt
/*
 * Copyright 2017-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
 */

package kotlinx.kover.features.jvm.impl

import com.intellij.rt.coverage.aggregate.api.AggregatorApi
import com.intellij.rt.coverage.aggregate.api.Request
import com.intellij.rt.coverage.report.api.Filters
import com.intellij.rt.coverage.verify.Verifier
import com.intellij.rt.coverage.verify.api.Counter
import com.intellij.rt.coverage.verify.api.Target
import com.intellij.rt.coverage.verify.api.ValueType
import com.intellij.rt.coverage.verify.api.VerificationApi
import kotlinx.kover.features.jvm.*
import java.io.File
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.*

private typealias IntellijRule = com.intellij.rt.coverage.verify.api.Rule

private typealias IntellijBound = com.intellij.rt.coverage.verify.api.Bound

internal object LegacyVerification {
    internal val ONE_HUNDRED: BigDecimal = BigDecimal(100)

    fun verify(
        rules: List<Rule>,
        tempDir: File,
        filters: ClassFilters,
        reports: List<File>,
        outputs: List<File>
    ): List<RuleViolations> {
        val intellijFilters: Filters = filters.convert()

        val rulesArray = ArrayList<IntellijRule>()

        val ic = File(tempDir, "agg-ic.ic")
        val smap = File(tempDir, "agg-smap.smap")

        val requests = Request(intellijFilters, ic, smap)
        AggregatorApi.aggregate(listOf(requests), reports, outputs)

        for (ruleIndex in rules.indices) {
            val rule = rules[ruleIndex]

            val bounds: MutableList<IntellijBound> = ArrayList()
            for (boundIndex in rule.bounds.indices) {
                val b = rule.bounds[boundIndex]

                bounds.add(
                    IntellijBound(
                        boundIndex,
                        counterToIntellij(b),
                        valueTypeToIntellij(b),
                        valueToIntellij(b, b.minValue),
                        valueToIntellij(b, b.maxValue)
                    )
                )
            }

            rulesArray.add(IntellijRule(ruleIndex, ic, targetToIntellij(rule), bounds))
        }


        val verifier = Verifier(rulesArray)
        verifier.processRules()
        val violations = VerificationApi.verify(rulesArray)

        val ruleViolations = ArrayList<RuleViolations>()
        for (ruleViolation in violations) {
            // TreeMap is using for getting stable order in result List - in this case, it is easier to write tests and Gradle build cache will not miss
            val resultBounds = TreeMap<ViolationId, BoundViolation>()

            val rule = rules[ruleViolation.id]
            for (boundIndex in ruleViolation.violations.indices) {
                val boundViolation = ruleViolation.violations[boundIndex]
                val bound = rule.bounds[boundViolation.id]

                for (maxViolation in boundViolation.maxViolations) {
                    val entityName = if (rule.groupBy != GroupingBy.APPLICATION) maxViolation.targetName else null
                    resultBounds[ViolationId(boundViolation.id, entityName)] =
                        BoundViolation(bound, true, intellijToValue(maxViolation.targetValue, bound), entityName)
                }
                for (minViolation in boundViolation.minViolations) {
                    val entityName = if (rule.groupBy != GroupingBy.APPLICATION) minViolation.targetName else null
                    resultBounds[ViolationId(boundViolation.id, entityName)] =
                        BoundViolation(bound, false, intellijToValue(minViolation.targetValue, bound), entityName)
                }
            }

            ruleViolations.add(RuleViolations(rule, ArrayList(resultBounds.values)))
        }

        return ruleViolations
    }

    private fun intellijToValue(intellijValue: BigDecimal, bound: Bound): BigDecimal {
        return if (isPercentage(bound.aggregationForGroup)) {
            intellijValue.multiply(ONE_HUNDRED)
        } else {
            intellijValue
        }
    }

    private fun targetToIntellij(rule: Rule): Target? {
        return when (rule.groupBy) {
            GroupingBy.APPLICATION -> Target.ALL
            GroupingBy.CLASS -> Target.CLASS
            GroupingBy.PACKAGE -> Target.PACKAGE
        }
    }

    private fun counterToIntellij(bound: Bound): Counter? {
        return when (bound.coverageUnits) {
            CoverageUnit.LINE -> Counter.LINE
            CoverageUnit.INSTRUCTION -> Counter.INSTRUCTION
            CoverageUnit.BRANCH -> Counter.BRANCH
        }
    }

    private fun valueTypeToIntellij(bound: Bound): ValueType? {
        return when (bound.aggregationForGroup) {
            AggregationType.COVERED_COUNT -> ValueType.COVERED
            AggregationType.MISSED_COUNT -> ValueType.MISSED
            AggregationType.COVERED_PERCENTAGE -> ValueType.COVERED_RATE
            AggregationType.MISSED_PERCENTAGE -> ValueType.MISSED_RATE
        }
    }

    private fun valueToIntellij(bound: Bound, value: BigDecimal?): BigDecimal? {
        if (value == null) return null

        return if (isPercentage(bound.aggregationForGroup)) {
            value.divide(ONE_HUNDRED, 6, RoundingMode.HALF_UP)
        } else {
            value
        }
    }

    private fun isPercentage(aggregationType: AggregationType): Boolean {
        return aggregationType == AggregationType.COVERED_PERCENTAGE || aggregationType == AggregationType.MISSED_PERCENTAGE
    }

    private data class ViolationId(private val index: Int, private val entityName: String?) : Comparable<ViolationId> {
        override fun compareTo(other: ViolationId): Int {
            // first compared by index
            index.compareTo(other.index).takeIf { it != 0 }?.let { return it }

            // if indexes are equals then compare by entity name
            if (entityName == null) {
                // bounds with empty entity names goes first
                return if (other.entityName == null) 0 else -1
            }

            if (other.entityName == null) return 1

            return entityName.compareTo(other.entityName)
        }
    }
}
